pkg/operator/csi/csistorageclasscontroller/csi_storageclass_controller.go

```diff
@@ -164,3 +178,54 @@ func removeConditionFn(condType string) v1helpers.UpdateStatusFunc {
 		return nil
 	}
 }
+
+// StorageClassStateEvaluator evaluates the StorageClassState in the corresponding
+// ClusterCSIDriver and reconciles the StorageClass according to that policy.
+// If Managed, apply the SC. If Unmanaged, do nothing. If Removed, delete the SC.
+type StorageClassStateEvaluator struct {
+	kubeClient             kubernetes.Interface
+	clusterCSIDriverLister oplisters.ClusterCSIDriverLister
+	operatorClient         v1helpers.OperatorClient
+	eventRecorder          events.Recorder
+}
+
+func NewStorageClassStateEvaluator(
+	kubeClient kubernetes.Interface,
+	clusterCSIDriverLister oplisters.ClusterCSIDriverLister,
+	eventRecorder events.Recorder) *StorageClassStateEvaluator {
+	return &StorageClassStateEvaluator{
+		kubeClient:             kubeClient,
+		clusterCSIDriverLister: clusterCSIDriverLister,
+		eventRecorder:          eventRecorder,
+	}
+}
+
+func (e *StorageClassStateEvaluator) Eval(ctx context.Context, expectedSC *storagev1.StorageClass) error {
+	// Look for the corresponding ClusterCSIDriver, but if this fails
+	// then assume the default "Managed" policy.
+	scState := operatorapi.ManagedStorageClass
+	clusterCSIDriver, err := e.clusterCSIDriverLister.Get(expectedSC.Provisioner)
+	if err != nil {
+		klog.V(4).Infof("failed to get ClusterCSIDriver %s, assuming Managed StorageClassState: %w", expectedSC.Provisioner, err)
+	} else {
+		scState = clusterCSIDriver.Spec.StorageClassState
+	}
+	// StorageClassState determines how the SC is reconciled.
+	switch scState {
+	case "":
+		// empty value defaults to managed policy
+		fallthrough
+	case operatorapi.ManagedStorageClass:
```

Comment:
nit: can this be written as - `case operatorapi.ManagedStorageClass, "":` rather than the fallthrough?

Reply:
Yeah I like that, updated.